discovery-provider/src/tasks/user_replica_set.py

```diff
@@ -0,0 +1,269 @@
+# Attempt to retrieve from cached values populated in index_network_peers.py
+# If unavailable, then a fallback to ethereum mainnet contracts will occur
+# Note that in the case of an invalid spID - one that is not yet registered on
+# the ethereum mainnet contracts, there will be an empty value in the returned
+# creator_node_endpoint
+# If this discrepancy occurs, a client replica set health check sweep will
+# result in a client-initiated failover operation to a valid set of replicas
+def get_endpoint_string_from_sp_ids(
+        self,
+        update_task,
+        primary,
+        secondaries,
+        redis
+):
+    sp_factory_inst = None
+    endpoint_string = None
+    primary_endpoint = None
+    try:
+        sp_factory_inst, primary_endpoint = get_endpoint_from_id(
+            self,
+            update_task,
+            sp_factory_inst,
+            primary
+        )
+        endpoint_string = "{}".format(primary_endpoint)
+        for secondary_id in secondaries:
+            secondary_endpoint = None
+            sp_factory_inst, secondary_endpoint = get_endpoint_from_id(
+                self,
+                update_task,
+                sp_factory_inst,
+                secondary_id
+            )
+            if secondary_endpoint:
+                endpoint_string = "{},{}".format(endpoint_string, secondary_endpoint)
+            else:
+                logger.info(f"user_replica_set.py | Failed to find secondary info for {secondary_endpoint}")
+    except Exception as exc:
+        logger.error(f"user_replica_set.py | ERROR in get_endpoint_string_from_sp_ids {exc}")
+    logger.info(f"user_replica_set.py | constructed {endpoint_string} from {primary},{secondaries}")
+    return endpoint_string
+
+# Initializes sp_factory if necessary and retrieves spID
+# Returns initialized instance of contract and endpoint
+def get_endpoint_from_id(self, update_task, sp_factory_inst, sp_id):
+    endpoint = None
+    # Get sp_id cache key
+    cache_key = get_sp_id_key(sp_id)
+    # Attempt to fetch from cache
+    sp_info_cached = get_pickled_key(update_task.redis, cache_key)
+    if sp_info_cached:
+        endpoint = sp_info_cached[1]
+        logger.info(f"user_replica_set.py | CACHE HIT FOR {cache_key}, found {sp_info_cached}")
+        return sp_factory_inst, endpoint
+
+    if not endpoint:
+        logger.info(f"user_replica_set.py | CACHE MISS FOR {cache_key}, found {sp_info_cached}")
+        if sp_factory_inst is None:
+            sp_factory_inst = get_sp_factory_inst(self, update_task)
+
+        cn_endpoint_info = sp_factory_inst.functions.getServiceEndpointInfo(
+            content_node_service_type,
+            sp_id
+        ).call()
+        logger.info(cn_endpoint_info)
+        endpoint = cn_endpoint_info[1]
+
+    return sp_factory_inst, endpoint
+
+# Return instance of ServiceProviderFactory initialized with configs
+def get_sp_factory_inst(self, update_task):
```

Comment:
nit - looks like this is shared with index_network_peers, could probably abstract it out

Reply:
wont fix, i think it's better to keep this here